In [ ]:
#importing relevant libraries for analysis and visualization
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings("ignore")

### **1.	What would be the suggested steps to make this Supervised Learning Model?**
*   First and most important thing to with the data is the cleaning part. In this part we import the data from the CSV file into the     notebook and remove all the non-ASCII characters.
*   First and most important thing to with the data is the cleaning part. In this part we import the data from the CSV file into the notebook and remove all the non-ASCII characters.




### **3.	Do you consider a need to apply any preprocessing on the training dataset? If so, why?**
######**Preprocesing starts after this Text block* 


*   Removed non-ASCII values from the **category of expenses** column.
*   converted date object to **dateTime** format for their use in analysis. 
*   Converting the dtype of **amount** from object to **float** and removed unnecessary characters like "-", replaced "," by "." to make sure the data transforms into float.
* Created **dummy variables** for the gender column and created a **new feature male** and removed the old gender feature. This was done because ML models expects data to be in numeric form.







###**Preprocessing**

In [ ]:
#importing the semi-colon seperated data
df = pd.read_csv("base_poupacerto_20200519.csv", encoding = 'latin', sep=';')
df.head()     #Top 5 rows in the data

id  ... pais_estabelecimento,,,
0  4,52863E+11  ...                      ,,
1  4,52863E+11  ...                     BR,
2  4,52863E+11  ...                      ,,
3  4,52863E+11  ...                     BR,
4  4,52863E+11  ...                     BR,

[5 rows x 14 columns]

In [ ]:
df.id.nunique()

29

In [ ]:
#checking the null values
df.isnull().sum()

id                          0
celular                     0
safra_abertura              0
cidade                      0
estado                      0
idade                       0
sexo                        0
limite_total                0
limite_disp                 0
data                        0
 valor                      0
grupo_estabelecimento       0
cidade_estabelecimento     19
pais_estabelecimento,,,    12
dtype: int64

*   *City and Country has some missing values (Comparitively lower than the whole data)*









In [ ]:
df.shape

(4955, 14)



*   *Total 4955 rows in the data with 14 columns (features)*






In [ ]:
#unique values in the columns
df.nunique(axis=0)

id                           29
celular                      29
safra_abertura               23
cidade                       10
estado                        2
idade                        20
sexo                          2
limite_total                 55
limite_disp                3801
data                        403
 valor                     2710
grupo_estabelecimento        25
cidade_estabelecimento      141
pais_estabelecimento,,,      14
dtype: int64

In [ ]:
#columns in the data
print(df.columns)
#converting the columns into Enlish Language for better understanding
df.columns = ['id', 'mob_no', 'branch_no', 'city', 'state', 'age', 'gender',
              'total_CC_limit', 'current_avl_limit', 'date', 'amount', 'categ_exp', 'purchasing_city', 'purchasing_country']

Index(['id', 'celular', 'safra_abertura', 'cidade', 'estado', 'idade', 'sexo',
       'limite_total', 'limite_disp', 'data', ' valor ',
       'grupo_estabelecimento', 'cidade_estabelecimento',
       'pais_estabelecimento,,,'],
      dtype='object')


In [ ]:
df.head()

id               mob_no  ...  purchasing_city purchasing_country
0  4,52863E+11  +55 (11) 96387-9924  ...        SAO PAULO                 ,,
1  4,52863E+11  +55 (11) 96387-9924  ...           SANTOS                BR,
2  4,52863E+11  +55 (11) 96387-9924  ...        SAO PAULO                 ,,
3  4,52863E+11  +55 (11) 96387-9924  ...           OSASCO                BR,
4  4,52863E+11  +55 (11) 96387-9924  ...           OSASCO                BR,

[5 rows x 14 columns]

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4955 entries, 0 to 4954
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   id                  4955 non-null   object
 1   mob_no              4955 non-null   object
 2   branch_no           4955 non-null   int64 
 3   city                4955 non-null   object
 4   state               4955 non-null   object
 5   age                 4955 non-null   int64 
 6   gender              4955 non-null   object
 7   total_CC_limit      4955 non-null   int64 
 8   current_avl_limit   4955 non-null   int64 
 9   date                4955 non-null   object
 10  amount              4955 non-null   object
 11  categ_exp           4955 non-null   object
 12  purchasing_city     4936 non-null   object
 13  purchasing_country  4943 non-null   object
dtypes: int64(4), object(10)
memory usage: 542.1+ KB


In [ ]:
df.date

0       04/12/19
1       09/11/19
2       06/05/19
3       01/06/19
4       01/06/19
          ...   
4950    23/01/20
4951    23/01/20
4952    19/02/20
4953    30/03/20
4954    24/01/20
Name: date, Length: 4955, dtype: object

In [ ]:
#converting date to datetime format
df.date = pd.to_datetime(df.date)

In [ ]:
df.date

0      2019-04-12
1      2019-09-11
2      2019-06-05
3      2019-01-06
4      2019-01-06
          ...    
4950   2020-01-23
4951   2020-01-23
4952   2020-02-19
4953   2020-03-30
4954   2020-01-24
Name: date, Length: 4955, dtype: datetime64[ns]

In [ ]:
df.head()

id               mob_no  ...  purchasing_city purchasing_country
0  4,52863E+11  +55 (11) 96387-9924  ...        SAO PAULO                 ,,
1  4,52863E+11  +55 (11) 96387-9924  ...           SANTOS                BR,
2  4,52863E+11  +55 (11) 96387-9924  ...        SAO PAULO                 ,,
3  4,52863E+11  +55 (11) 96387-9924  ...           OSASCO                BR,
4  4,52863E+11  +55 (11) 96387-9924  ...           OSASCO                BR,

[5 rows x 14 columns]

In [ ]:
df.amount = df.amount.apply(lambda x : x.replace(',', '.').replace('.', ''))  #replacing comma with decimal, removing unnnecessary full stop
i = df[df.amount==' -   '].index     # index of non-specified value in amount columns (-), needed to be removed
df.drop(i, inplace=True)
df.amount = df.amount.astype(float)      #converting amount to float

In [ ]:
#converted amount to float and date to dateTime format
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4942 entries, 0 to 4954
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   id                  4942 non-null   object        
 1   mob_no              4942 non-null   object        
 2   branch_no           4942 non-null   int64         
 3   city                4942 non-null   object        
 4   state               4942 non-null   object        
 5   age                 4942 non-null   int64         
 6   gender              4942 non-null   object        
 7   total_CC_limit      4942 non-null   int64         
 8   current_avl_limit   4942 non-null   int64         
 9   date                4942 non-null   datetime64[ns]
 10  amount              4942 non-null   float64       
 11  categ_exp           4942 non-null   object        
 12  purchasing_city     4935 non-null   object        
 13  purchasing_country  4931 non-null   object      

In [ ]:
#converting gender to dummy variables
dummy = pd.get_dummies(df.gender, drop_first=True)       #removed gender feature and replaced with male (1 for male, 0 for female) 
df = pd.concat([df, dummy], axis = 1)
df.drop('gender', axis = 1, inplace=True)
df.rename(columns = {'M':'male'}, inplace = True) 

In [ ]:
df.head()

id               mob_no  ...  purchasing_country male
0  4,52863E+11  +55 (11) 96387-9924  ...                  ,,    0
1  4,52863E+11  +55 (11) 96387-9924  ...                 BR,    0
2  4,52863E+11  +55 (11) 96387-9924  ...                  ,,    0
3  4,52863E+11  +55 (11) 96387-9924  ...                 BR,    0
4  4,52863E+11  +55 (11) 96387-9924  ...                 BR,    0

[5 rows x 14 columns]